### PR TITLE
[stable] std.process: Include program name in exception msg if it cannot be found

### PR DESCRIPTION
**Pull request by kinke**
On Windows, the message for this common error case wasn't as informative as it should be:
```
Failed to spawn new process (The system cannot find the file specified.)
```

Handle this just like on POSIX via a message like:
```
Executable file not found: 7z.exe
```

### CODE REVIEW

**Review by thewilsonator — APPROVED** [review]

File: std/process.d
Comment (CyberShadow) [inline std/process.d:1296]: What happens if a DLL or other resource is missing?

Comment (kinke) [reply std/process.d:1296]: For a missing DLL, it doesn't seem to result in CreateProcessW failing (and instead make the child process exit with != 0). Tested with a `pythonw.exe` and taking away its `python37.dll`.

**Comment by CyberShadow** [comment]
I think https://github.com/dlang/phobos/pull/7394 aimed for the same goal, but at the end was a smaller diff and made fewer assumptions. The author lost interest, perhaps it should be continued instead?

**Comment by kinke** [comment]
I don't have a preference wrt. extra struct vs. extra param + some more versioning lines - I only care about the msg, as pretty much all ProcessExceptions I've ever seen on Windows were due to some missing program or incomplete PATH, and all I ever wanted to know was what it was trying to invoke. Using the program as additional context for all other CreateProcess errors might be worth it; that was actually my first version, before I checked what happens on Posix, after which I went for a simpler msg consistent with the Posix one.

**Comment by CyberShadow** [comment]
If we can make things both useful and consistent across platforms, I think that would be a win :)

**Comment by kinke** [comment]
On Posix, that msg is used when the program is only a name (no dir separator). With a path, it's `Not an executable file: ./iDontExist`. At least according to run.dlang.io - I'll need to debug why the testcase fails on Posix.

**Comment by thewilsonator** [comment]
please ping me when this is good to go.

**Comment by kinke** [comment]
Good now IMO, improving the msg on Posix too if the program contains a slash.

File: std/process.d
Comment (CyberShadow) [inline std/process.d:1172]: Because POSIX file names may contain characters such as newlines or quote characters, may I suggest
```suggestion
                    errorMsg = "Failed to execute " ~ escapeShellFileName(name);
```

File: std/process.d
Comment (CyberShadow) [inline std/process.d:1296]: I think the other PR was still a little better in this regard for a few reasons:

1. The error is "**file** not found", not "**executable** not found"
2. We forego error message localization provided by the OS for this error code only
3. It seems inconsistent to include the program name only for this error code, and not others.

**Review by CyberShadow — APPROVED** [review]

File: std/process.d
Comment (kinke) [reply std/process.d:1296]: Well, the other PR was abandoned, resulting in no progress at all. May I suggest you follow up on your own, given that you have a strong opinion about this and seem to care about the improbable error cases and Posix filenames as well? I'm thankful for concerns regarding the DLLs, those haven't crossed my mind at all, but seem to be of no concern.

Comment (CyberShadow) [reply std/process.d:1296]: No problem. I don't really have a strong opinion on this, just pointing out a few opportunities for improvement while we're on the subject.

**Comment by CyberShadow** [comment]
BTW, why did this target stable?

File: std/process.d
Comment (CyberShadow) [inline std/process.d:1296]: Just noticed that this error message has the exact same wording as for when `searchPathFor` failed. We should probably avoid having errors that share the message string but indicate different error conditions.

**Comment by kinke** [comment]
> BTW, why did this target stable?

Because I hate waiting for months to finally see a trivial improvement such as this (targeting master now means it shows up in 2.096).

File: std/process.d
Comment (kinke) [reply std/process.d:1296]: Yeah, that's what I meant by 'consistent with the Posix msg' (`searchPathFor` is Posix only).

**Comment by CyberShadow** [comment]
I understand, but there are a lot of other PRs that this would apply to, and if that were to happen, I think that would considerably diminish the value of the stable branch.

There is also the matter that follow-up PRs need to target stable too (see #7727).

Perhaps consider using Digger to easily use the current DMD master? :)

File: std/process.d
Comment (CyberShadow) [reply std/process.d:1296]: OK. Sorry, at the time I did not know what exactly you meant by that. I understand that the two can refer to different error conditions (a file name was not found in `$PATH` vs. an absolute file name does not exist on disk).

Comment (kinke) [reply std/process.d:1296]: Can be relative too. FWIW, I think Posix resulting in 2 different msges for the same error (program not found) isn't ideal; to me it doesn't matter whether the program contains a slash or not.

Comment (CyberShadow) [reply std/process.d:1296]: > to me it doesn't matter whether the program contains a slash or not.

What do you mean? Whether the path contains a path separator or not is what decides if the `PATH` variable is consulted. The `Posix` logic mirrors the Windows API logic (quote "If the file name does not contain a directory path"), to make the Phobos API consistent across both platforms.

**Comment by kinke** [comment]
> There is also the matter that follow-up PRs need to target stable too

Yes, or wait until stable is merged into master again.

> Perhaps consider using Digger to easily use the current DMD master? :)

This was bothering me in production (countless errors due to missing programs and me having no idea what it tried to invoke), where using DMD master is out of the question.

**Comment by CyberShadow** [comment]
> This was bothering me in production (countless errors due to missing programs and me having no idea what it tried to invoke),

I see. I would use Process Monitor for this case, but I'm guessing that wasn't an option.

> where using DMD master is out of the question.

FWIW, Digger also allows cherry-picking changes, e.g. `stable+phobos#7726`.

**Comment by CyberShadow** [comment]
Another reason to not put things in stable unless necessary: https://github.com/dlang/phobos/pull/7727#discussion_r546227580

@thewilsonator 👆

**Pull request merged**
